### ISSUE TITLE
GPU utilization is low

### ISSUE THREAD

**Issue by ggaemo** [facebookresearch/pythia#210]
## ❓ Questions and Help

Hello I'm running the vqa2 script with the following python file and arguments

 python tools/run.py --tasks vqa --datasets vqa2 --model pythia --config configs/vqa/vqa2/pythia_train_and_val.yml --resume_file d
ata/models/pythia_train_val.pth --evalai_inference 1 --run_type inference

however, it seems that the gpu utilization is low. I've seen in #170 that you've got an average of 98% gpu utilization. However, when I run the script above, it seems that there are bursts of 100% utilization for a brief second, and it remains idle for most of the time. The total inference time shows more than 3 hours. 

I'm running this on a single Titan X GPU with CPUs with 56 cores, so hardware is not a bottleneck here.

Another warning that I'm getting is 

 UserWarning: 'optimizer' key is not present in the checkpoint asked to be loaded. Skipping
.
  "'optimizer' key is not present in the "

Is this the way it should be? I think that the checkpoint file should have the optimizer for further training if I'm using this model for pretrained models.

**Comment by ggaemo** [comment]
![image](https://user-images.githubusercontent.com/8081512/71144488-85316380-2261-11ea-9571-37b12f005a73.png)


it seems that the data workers are not doing enough jobs. The running time of the data workers are far too low.

**Comment by apsdehal** [comment]
This is a known issue on which we are working on and would be improved in v0.4 release. You can move to the v0.4 to use distributed training setup which should help improve on the GPU utilization. Also, set the `training_parameters.pin_memory` to true. 

For the optimizer, we only provide pretrained models as of now, so we didn't provide exact checkpoints to avoid any confusion. I can provide the checkpoint if you need it.

**Comment by ggaemo** [comment]
I tried the v0.4 

with the same argument and it gave an error like this.

![image](https://user-images.githubusercontent.com/8081512/71307963-8917de00-2439-11ea-9b3d-920d7f3ef80e.png)


So I checked the model_state_dict of the model that and it showed that it had the following keys.

odict_keys(['word_embedding.weight', 'text_embeddings.0.module.recurrent_unit.weight_ih_l0', 'text_embeddings.0.module.recurrent_unit.weight_hh_l0', 'text_embeddings.0.module.recurrent_unit.bias_ih_l0', 'text_embeddings.0.module.recurrent_unit.bias_hh_l0', 'text_embeddings.0.module.conv1.weight', 'text_embeddings.0.module.conv1.bias', 'text_embeddings.0.module.conv2.weight', 'text_embeddings.0.module.conv2.bias', 'image_feature_encoders.0.module.lc.weight', 'image_feature_encoders.0.module.lc.bias', 'image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.bias', 'image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_g', 'image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_v', 'image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.bias', 'image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_g', 'image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_v', 'image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.bias', 'image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.weight_g', 'image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.weight_v', 'image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.bias', 'image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_g', 'image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_v', 'image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.bias', 'image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_g', 'image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_v', 'image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.bias', 'image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.weight_g', 'image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.weight_v', 'image_text_multi_modal_combine_layer.module.fa_image.layers.0.bias', 'image_text_multi_modal_combine_layer.module.fa_image.layers.0.weight_g', 'image_text_multi_modal_combine_layer.module.fa_image.layers.0.weight_v', 'image_text_multi_modal_combine_layer.module.fa_txt.layers.0.bias', 'image_text_multi_modal_combine_layer.module.fa_txt.layers.0.weight_g', 'image_text_multi_modal_combine_layer.module.fa_txt.layers.0.weight_v', 'classifier.module.f_o_text.layers.0.bias', 'classifier.module.f_o_text.layers.0.weight_g', 'classifier.module.f_o_text.layers.0.weight_v', 'classifier.module.f_o_image.layers.0.bias', 'classifier.module.f_o_image.layers.0.weight_g', 'classifier.module.f_o_image.layers.0.weight_v', 'classifier.module.linear_text.weight', 'classifier.module.linear_text.bias', 'classifier.module.linear_image.weight', 'classifier.module.linear_image.bias'])


Maybe the the name "fa_context.layers" is changed to "fa_txt.layers" is this right?

**Comment by ggaemo** [comment]
I looked into it for a better comparison

the following are from the checkpoint file of "pythia.pth" it shows the parameter name and size.


module.word_embedding.weight torch.Size([75505, 300])
module.text_embeddings.0.module.recurrent_unit.weight_ih_l0 torch.Size([4096, 300])
module.text_embeddings.0.module.recurrent_unit.weight_hh_l0 torch.Size([4096, 1024])
module.text_embeddings.0.module.recurrent_unit.bias_ih_l0 torch.Size([4096])
module.text_embeddings.0.module.recurrent_unit.bias_hh_l0 torch.Size([4096])
module.text_embeddings.0.module.conv1.weight torch.Size([512, 1024, 1])
module.text_embeddings.0.module.conv1.bias torch.Size([512])
module.text_embeddings.0.module.conv2.weight torch.Size([2, 512, 1])
module.text_embeddings.0.module.conv2.bias torch.Size([2])
module.image_feature_encoders.0.module.lc.weight torch.Size([2048, 2048])
module.image_feature_encoders.0.module.lc.bias torch.Size([2048])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.bias torch.Size([5000])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_g torch.Size([])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_v torch.Size([5000, 2048])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.bias torch.Size([5000])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_g torch.Size([])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_v torch.Size([5000, 2048])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.bias torch.Size([5000])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_g torch.Size([])
module.image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_v torch.Size([5000, 2048])
module.image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.bias torch.Size([1])
module.image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.weight_g torch.Size([])
module.image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.weight_v torch.Size([1, 5000])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.bias torch.Size([5000])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_g torch.Size([])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_v torch.Size([5000, 2048])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.bias torch.Size([5000])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_g torch.Size([])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_v torch.Size([5000, 2048])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.bias torch.Size([5000])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_g torch.Size([])
module.image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_v torch.Size([5000, 2048])
module.image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.bias torch.Size([1])
module.image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.weight_g torch.Size([])
module.image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.weight_v torch.Size([1, 5000])
module.image_text_multi_modal_combine_layer.module.fa_image.layers.0.bias torch.Size([5000])
module.image_text_multi_modal_combine_layer.module.fa_image.layers.0.weight_g torch.Size([])
module.image_text_multi_modal_combine_layer.module.fa_image.layers.0.weight_v torch.Size([5000, 4096])
module.image_text_multi_modal_combine_layer.module.fa_txt.layers.0.bias torch.Size([5000])
module.image_text_multi_modal_combine_layer.module.fa_txt.layers.0.weight_g torch.Size([])
module.image_text_multi_modal_combine_layer.module.fa_txt.layers.0.weight_v torch.Size([5000, 2048])
module.image_text_multi_modal_combine_layer.module.fa_context.layers.0.bias torch.Size([5000])
module.image_text_multi_modal_combine_layer.module.fa_context.layers.0.weight_g torch.Size([])
module.image_text_multi_modal_combine_layer.module.fa_context.layers.0.weight_v torch.Size([5000, 2048])
module.classifier.module.f_o_text.layers.0.bias torch.Size([300])
module.classifier.module.f_o_text.layers.0.weight_g torch.Size([])
module.classifier.module.f_o_text.layers.0.weight_v torch.Size([300, 5000])
module.classifier.module.f_o_image.layers.0.bias torch.Size([5000])
module.classifier.module.f_o_image.layers.0.weight_g torch.Size([])
module.classifier.module.f_o_image.layers.0.weight_v torch.Size([5000, 5000])
module.classifier.module.linear_text.weight torch.Size([3129, 300])
module.classifier.module.linear_text.bias torch.Size([3129])
module.classifier.module.linear_image.weight torch.Size([3129, 5000])
module.classifier.module.linear_image.bias torch.Size([3129])

The following are printed from the model loaded by the yaml file "pythia_train_and_val.yml".
text_embeddings.0.module.recurrent_unit.weight_ih_l0 torch.Size([4096, 300])
text_embeddings.0.module.recurrent_unit.weight_hh_l0 torch.Size([4096, 1024])
text_embeddings.0.module.recurrent_unit.bias_ih_l0 torch.Size([4096])
text_embeddings.0.module.recurrent_unit.bias_hh_l0 torch.Size([4096])
text_embeddings.0.module.conv1.weight torch.Size([512, 1024, 1])
text_embeddings.0.module.conv1.bias torch.Size([512])
text_embeddings.0.module.conv2.weight torch.Size([2, 512, 1])
text_embeddings.0.module.conv2.bias torch.Size([2])
image_feature_encoders.0.module.lc.weight torch.Size([2048, 2048])
image_feature_encoders.0.module.lc.bias torch.Size([2048])
image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.bias torch.Size([5000])
image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_g torch.Size([])
image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_v torch.Size([5000, 2048])
image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.bias torch.Size([5000])
image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_g torch.Size([])
image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_v torch.Size([5000, 2048])
image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.bias torch.Size([1])
image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.weight_g torch.Size([])
image_feature_embeddings_list.0.0.image_attention_model.module.transform.module.lc.weight_v torch.Size([1, 5000])
image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.bias torch.Size([5000])
image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_g torch.Size([])
image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_image.layers.0.weight_v torch.Size([5000, 2048])
image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.bias torch.Size([5000])
image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_g torch.Size([])
image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_txt.layers.0.weight_v torch.Size([5000, 2048])
image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.bias torch.Size([1])
image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.weight_g torch.Size([])
image_feature_embeddings_list.1.0.image_attention_model.module.transform.module.lc.weight_v torch.Size([1, 5000])
image_text_multi_modal_combine_layer.module.fa_image.layers.0.bias torch.Size([5000])
image_text_multi_modal_combine_layer.module.fa_image.layers.0.weight_g torch.Size([])
image_text_multi_modal_combine_layer.module.fa_image.layers.0.weight_v torch.Size([5000, 4096])
image_text_multi_modal_combine_layer.module.fa_txt.layers.0.bias torch.Size([5000])
image_text_multi_modal_combine_layer.module.fa_txt.layers.0.weight_g torch.Size([])
image_text_multi_modal_combine_layer.module.fa_txt.layers.0.weight_v torch.Size([5000, 2048])
classifier.module.f_o_text.layers.0.bias torch.Size([300])
classifier.module.f_o_text.layers.0.weight_g torch.Size([])
classifier.module.f_o_text.layers.0.weight_v torch.Size([300, 5000])
classifier.module.f_o_image.layers.0.bias torch.Size([5000])
classifier.module.f_o_image.layers.0.weight_g torch.Size([])
classifier.module.f_o_image.layers.0.weight_v torch.Size([5000, 5000])
classifier.module.linear_text.weight torch.Size([3129, 300])
classifier.module.linear_text.bias torch.Size([3129])
classifier.module.linear_image.weight torch.Size([3129, 5000])
classifier.module.linear_image.bias torch.Size([3129])

It seems that the model from v0.4 has an additional context layer. Is this mentioned in the paper?

**Comment by apsdehal** [comment]
Can you pull the latest commit in v0.4 and let me know if it works?

**Comment by ggaemo** [comment]
![image](https://user-images.githubusercontent.com/8081512/71328891-c8d9e500-2561-11ea-8ee6-5a76ade204f3.png)

I looked at the latest commit and pulled it. Still, the **fa_context** components are missing.

![image](https://user-images.githubusercontent.com/8081512/71328903-f7f05680-2561-11ea-870f-ce8d3dac196c.png)

The image above is the Image Encoder part of the whole model. The left column shows the model from v0.3(master branch), and the right column shows the model from v0.4 branch. The difference is the fa_context layer that is in the same hierarchy as the fa_image, fa_txt layer inside the "combination layer" which the v0.4 branch version does not have.
The pretrained model also has the fa_context layer which seems like the reason why this is happening.

If you run the code, the following error message appears, which all shows that parameters with 'fa_context' included is not present in the v0.4 branch model.

Unexpected key(s) in state_dict: "image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.bias", "image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_g", "image_feature_embeddings_list.0.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_v", "image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.bias", "image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_g", "image_feature_embeddings_list.1.0.image_attention_model.module.combination_layer.module.fa_context.layers.0.weight_v", "image_text_multi_modal_combine_layer.module.fa_context.layers.0.bias", "image_text_multi_modal_combine_layer.module.fa_context.layers.0.weight_g", "image_text_multi_modal_combine_layer.module.fa_context.layers.0.weight_v".

**Comment by apsdehal** [comment]
Can you change: https://github.com/facebookresearch/pythia/blob/v0.4/pythia/utils/checkpoint.py#L118 to `self.trainer.model.load_state_dict(final_dict, strict=False)`.

**Comment by ggaemo** [comment]
Thanks, now it works. If it is not to bothersome for you, could you explain what the fa_context layers were doing? fa_image and fa_text was mentioned in the paper for merging the two modalities, but I don't see an additional layer that has a context name in it. 

Also I've mentioned this in another issue but I'll ask it here as well: 
if I put the training_parameters.pin_memory to True, it runs to training mode even if I've designated run_type to val.

**Comment by apsdehal** [comment]
fa_context is used is lorra and was unused extra layer for pythia model. That's way the code was update to not load it in case of pythia. `pin_memory` issue needs to be figured out, please open up a separate task for it. For the optimizer, I only have checkpoint for the model trained on only train. Would that work?

**Comment by ggaemo** [comment]
Thanks, that checkpoint would be very helpful. I'll open up the pin_memory issue.
